src/dotnet/UI.Blazor/Services/PanelsUI/LeftPanel.cs

```diff
@@ -0,0 +1,75 @@
+using ActualChat.UI.Blazor.Services.Internal;
+
+namespace ActualChat.UI.Blazor.Services;
+
+public class LeftPanel
+{
+    private readonly IMutableState<bool> _isVisible;
+    private readonly object _lock = new();
+
+    private ILogger Log { get; }
+    public PanelsUI Owner { get; }
+    // ReSharper disable once InconsistentlySynchronizedField
+    public IState<bool> IsVisible => _isVisible;
+    public event EventHandler? VisibilityChanged;
+
+    public LeftPanel(PanelsUI owner)
+    {
+        Owner = owner;
+        Log = Owner.Services.LogFor(GetType());
+
+        _isVisible = Owner.Services.StateFactory().NewMutable(true);
+        Owner.History.Register(new OwnHistoryState(this, true));
+    }
+
+    public void SetIsVisible(bool value)
+    {
+        var localUrl = Owner.History.LocalUrl;
+        value |= localUrl.IsChatRoot(); // Always visible if @ /chat
+        value &= !localUrl.IsDocsOrDocsRoot(); // Always invisible if @ /docs*
+        value |= IsWide(); // Always visible if wide
+
+        bool oldIsVisible;
+        lock (_lock) {
+            oldIsVisible = _isVisible.Value;
+            if (oldIsVisible != value)
+                _isVisible.Value = value;
+        }
+        if (oldIsVisible != value) {
+            Log.LogDebug("Visibility changed: {IsVisible}", value);
+            Owner.History.Save<OwnHistoryState>();
```

Comment:
due to lock above you expect race conditions in this method.
On the other hand here you save to history state IsVisible.Value in a not concurrent safe manner. At this line you can safe state which is already modified by a concurrent call.
i.e. you can save `false, false` when you must save `false, true `

Reply:
kept it without changes